File: src/main/java/de/danielmaile/aether/item/ListenerBlock.java
package de.danielmaile.aether.item;

import de.danielmaile.aether.util.NBTEditor;
import de.danielmaile.aether.worldgen.AetherWorld;
import org.bukkit.Location;
import org.bukkit.block.Block;
import org.bukkit.block.data.BlockData;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.inventory.ItemStack;

public class ListenerBlock implements Listener
{
    @EventHandler
    public void onBlockBreak(BlockBreakEvent event)
    {
        if (event.isCancelled()) return;

        Block block = event.getBlock();
        if (!block.getWorld().equals(AetherWorld.getWorld())) return;

        Location location = block.getLocation();
        BlockType blockType = BlockType.getFromMaterial(block.getType());
        if (blockType == null) return;

        block.getWorld().dropItemNaturally(location, blockType.getItemDrop().getItemStack());
        event.setDropItems(false);
    }

    @EventHandler
    public void onBlockPlace(BlockPlaceEvent event)
    {
        if (event.isCancelled()) return;

        Block block = event.getBlock();
        ItemStack itemStack = event.getItemInHand();
        if (block.getWorld().equals(AetherWorld.getWorld()))
        {
            if (NBTEditor.hasKey(itemStack, ItemType.AETHER_ITEM_TAG_KEY))
            {
                //Convert Item to correct material
                ItemType itemType = ItemType.fromTag(itemStack);
                if (itemType != null)
                {
                    //Change material but keep orientation
                    BlockData data = block.getBlockData();
                    block.setType(itemType.getPlaceMaterial());
                    block.setBlockData(data);
                }
            }
            else
            {
                //Check if block which a player tries to place is already used by aether block
                BlockType blockType = BlockType.getFromMaterial(itemStack.getType());
                if (blockType != null)
                {
                    event.setCancelled(true);
                }
            }
        }
        else
        {
            //Aether items can't be placed in other worlds
            if (NBTEditor.hasKey(itemStack, ItemType.AETHER_ITEM_TAG_KEY))
            {
                event.setCancelled(true);
            }
        }
    }
}

File: src/main/java/de/danielmaile/lama/aether/item/ListenerBlock.kt
package de.danielmaile.lama.aether.item

import de.danielmaile.lama.aether.inst
import de.danielmaile.lama.aether.util.hasKey
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent

class ListenerBlock : Listener {

    @EventHandler
    fun onBlockBreak(event: BlockBreakEvent) {
        if (event.isCancelled) return

        val block = event.block
        if (block.world != inst().aetherWorld.getWorld()) return

        val location = block.location
        val blockType = BlockType.fromMaterial(block.type) ?: return

        block.world.dropItemNaturally(location, blockType.itemDrop.getItemStack())
        event.isDropItems = false
    }

    @EventHandler
    fun onBlockPlace(event: BlockPlaceEvent) {
        if (event.isCancelled) return

        val block = event.block
        val itemStack = event.itemInHand

        if (block.world == inst().aetherWorld.getWorld()) {
            if (itemStack.hasKey(AETHER_ITEM_TAG_KEY)) {
                //Convert Item to correct material
                val itemType = ItemType.fromTag(itemStack)
                if (itemType != null) {
                    //Change material but keep orientation
                    val data = block.blockData
                    block.type = itemType.placeMaterial!!
                    block.blockData = data
                }
            } else {
                //Check if block which a player tries to place is already used by aether block
                val blockType = BlockType.fromMaterial(itemStack.type)
                if (blockType != null) {
                    event.isCancelled = true
                }
            }
        } else {
            //Aether items can't be placed in other worlds
            if (itemStack.hasKey(AETHER_ITEM_TAG_KEY)) {
                event.isCancelled = true
            }
        }
    }
}